src/common/subprocess/subprocess.h

```diff
@@ -0,0 +1,370 @@
+ *  flux_cmd_add_channel().  If 'stream' is NULL, defaults to "STDIN".
+ *
+ *  Returns the total amount of data successfully buffered.
+ */
+int flux_subprocess_write (flux_subprocess_t *p, const char *stream,
+                           const char *buf, size_t len);
+
+/*
+ *  Close "stream" stream of subprocess `p` and schedule EOF to be sent.
+ *  'stream' can be "STDIN" or the name of a stream specified with
+ *  flux_cmd_add_channel().  If 'stream' is NULL, defaults to "STDIN".
+ */
+int flux_subprocess_close (flux_subprocess_t *p, const char *stream);
+
+/*
+ *  Read up to `len` bytes of unread data from stream `stream`.  To
+ *   read all data, specify 'len' of -1.  'stream' can be "STDOUT",
+ *   "STDERR", or the name of a stream specified with
+ *   flux_cmd_add_channel().  If 'stream' is NULL, defaults to
+ *   "STDOUT".
+ *
+ *   Returns pointer to buffer on success and NULL on error with errno
+ *   set.  If reading from "STDOUT" or "STDERR", buffer is guaranteed
+ *   to be NUL terminated.  User shall not free returned pointer.
+ *   Length of buffer returned can optionally returned in 'lenp'.  A
+ *   length of 0 indicates that the subprocess has closed this stream.
+ */
+const char *flux_subprocess_read (flux_subprocess_t *p,
+                                  const char *stream,
+                                  int len, int *lenp);
+
+/*
+ *  Read line unread data from stream `stream`.  'stream' can be
+ *   "STDOUT", "STDERR", or the name of a stream specified with
+ *   flux_cmd_add_channel().  If 'stream' is NULL, defaults to
+ *   "STDOUT".
+ *
+ *   Returns pointer to buffer on success and NULL on error with errno
+ *   set.  If reading from "STDOUT" or "STDERR", buffer is guaranteed
+ *   to be NUL terminated.  User shall not free returned pointer.
+ *   Length of buffer returned can optionally returned in 'lenp'.
+ */
+const char *flux_subprocess_read_line (flux_subprocess_t *p,
+                                       const char *stream,
+                                       int *lenp);
+
+/* Identical to flux_subprocess_read_line(), but does not return
+ * trailing newline.
+ */
+const char *flux_subprocess_read_trimmed_line (flux_subprocess_t *p,
+                                               const char *stream,
+                                               int *lenp);
+
+/*
+ *  Create RPC to send signal `signo` to subprocess `p`.
+ *  This call returns a flux_future_t. Use flux_future_then(3) to register
+ *   a continuation callback when the kill operation is complete, or
+ *   flux_future_wait_for(3) to block until the kill operation is complete.
+ */
+flux_future_t *flux_subprocess_kill (flux_subprocess_t *p, int signo);
+
+/*
+ *  Add/remove a reference to subprocess object `p`. The subprocess object
+ *   is destroyed once the last reference is removed.
+ */
+void flux_subprocess_ref (flux_subprocess_t *p);
+void flux_subprocess_unref (flux_subprocess_t *p);
+void flux_subprocess_destroy (void *arg);
```

Comment:
Should `flux_subprocess_destroy()` be exported via the public interface? Also, shouldn't it take `flux_subprocess_t *`?

Reply:
I specifically added it b/c I wanted to use ```flux_subprocess_unref()``` as a zhash/zlist free function.   But there was a typing issue b/c zhash/zlist needs a function with a ```void *``` parameter.

Perhaps this is a mistake and I should have just created a wrapper destroy function inside that code.

Reply:
Well, my main concern would just be that users of the API might be tempted to use `subprocess_destroy` after `create` instead of `unref` -- If the `destroy` function is moved to a "private" method, then the type of the argument is less important. If it needs to stay public, then I would suggest replacing the `void *` with `flux_subprocess_t *` and either wrapping the destroy function for zhash_t (or you can cast the function to the correct type)